### DISCUSSION TITLE
How to Implement Nodeselector for Kafka and Zookeeper

### DISCUSSION THREAD

**Discussion by jeffreywstevens** [strimzi/strimzi-kafka-operator#7522] [Q&A] [answered]
This works in my environment to keep pods from being deployed to Windows nodes.  
'  nodeSelector:
    kubernetes.io/os: linux'

How do I implement this for the kafka cluster yaml file for both kafka and zookeeper. When I tried affinity, it gives and errors, so I am doing something wrong, or I have outdated google hits.

I tried permutations of this but get validation errors.
`    template:
      pod:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                  - key: kubernetes.io/os
                    operator: In 
                    values:
                    - linux
`
[ValidationError(Kafka.spec.kafka): unknown field "affinity" in io.strimzi.kafka.v1beta2.Kafka.spec.kafka, 

I know more RTFM but does someone have a working example like this to help me out?

**Comment by jeffreywstevens** [comment, marked as answer]
I fixed it.

**Reply by scholzj** [reply]
Great. Maybe if you say how you fixed it, it could help others ;-)

**Reply by jeffreywstevens** [reply]
I did have one embarrassing typo.  Below is what worked. This also has to be applied to zookeeper and the operators in the Kafka object.

kafka:
    template:
      pod:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: kubernetes.io/os
                  operator: In
                  values:
                  - linux

**Comment by fayazahamedj** [comment]
Hello Everyone & @jeffreywstevens I tried to implement NodeSelector for both Kafka & zookeeper as suggested above by Jeffrey, but it's not taking effect. Could you take a look?

 ```
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  namespace: streaming-telemetry
spec:
   zookeeper:
      replicas: 3
      storage:
        size: 2Gi
        type: persistent-claim
      template:
         pod:
            affinity:
               nodeAffinity:
                  requiredDuringSchedulingIgnoredDuringExecution:
                     nodeSelectorTerms:
                        - matchExpressions:
                               - key: nodetype
                                  operator: In
                                  values:
                                  - kafka
